### ISSUE TITLE
send_nrdp.py client is not python 3

### ISSUE THREAD

**Issue by PHuhn** [NagiosEnterprises/nrdp#38]
I am working in python 3.  It would be nice if it was also available in python 3.
And also pl-lint.
Thanks..

**Comment by jomann09** [comment]
Thanks for reporting this, will look into making it work with all current versions of Python.

**Comment by PHuhn** [comment]
My version of python3 is as follows:
```
#!/usr/bin/env python3
#
# send_nrdp.py
#
# Copyright (c) 2010-2017 - Nagios Enterprises, LLC.
# Written by: Scott Wilkerson (nagios@nagios.org)
#
# 2017-09-22 Troy Lea aka BOX293
#  - Fixed setup function that was not working for piped results
#    as the xml string was missing "</checkresults>"
#  - Fixed setup function that was not working with arguments when
#    run as a cron job or if being used as a nagios command like
#    obsessive compulsive ... "if not sys.stdin.isatty():" was the
#    reason why.
# 2017-09-25 Troy Lea aka BOX293
#  - Added file argument to allow XML results to be read from file
#  - Replaced optparse with argparse so help text can include newline
#    characters, allowing for better formatting.
# 2017-09-26 Troy Lea aka BOX293
#  - Made sure url ends with a / before posting


import argparse, sys, html
from urllib import request as UrlRequest, parse as UrlParse
from xml.dom.minidom import parseString

class send_nrdp:
    def run(self):
        parser = argparse.ArgumentParser(formatter_class=argparse.RawTextHelpFormatter)
        
        parser.add_argument('-u', '--url', action="store",
            dest="url", help="\
** REQUIRED ** The URL used to access the remote NRDP agent.")
        parser.add_argument('-t', '--token', action="store",
            dest="token", help="\
** REQUIRED ** The authentication token used to access the\n\
remote NRDP agent.")
        parser.add_argument('-H', '--hostname', action="store",
            dest="hostname", help="\
The name of the host associated with the passive host/service\n\
check result.")
        parser.add_argument('-s', '--service', action="store",
            dest="service", help="\
For service checks, the name of the service associated with the\n\
passive check result.")
        parser.add_argument('-S', '--state', action="store",
            dest="state", help="\
An integer indicating the current state of the host or service.")
        parser.add_argument('-o', '--output', action="store",
            dest="output", help="\
Text output to be sent as the passive check result.\n\
Newlines should be encoded with encoded newlines (\\n).")
        parser.add_argument('-f', '--file', action="store",
            dest="file", help="\
This file will be sent to the NRDP server specified in -u\n\
The file should be an XML file in the following format:\n\
##################################################\n\
<?xml version='1.0'?>\n\
<checkresults>\n\
  <checkresult type=\"host\" checktype=\"1\">\n\
    <hostname>YOUR_HOSTNAME</hostname>\n\
    <state>0</state>\n\
    <output>OK|perfdata=1.00;5;10;0</output>\n\
  </checkresult>\n\
  <checkresult type=\"service\" checktype=\"1\">\n\
    <hostname>YOUR_HOSTNAME</hostname>\n\
    <servicename>YOUR_SERVICENAME</servicename>\n\
    <state>0</state>\n\
    <output>OK|perfdata=1.00;5;10;0</output>\n\
  </checkresult>\n\
</checkresults>\n\
##################################################")
        parser.add_argument('-d', '--delim', action="store",
            dest="delim", help="\
With only the required parameters send_nrdp.py is capable of\n\
processing data piped to it either from a file or other process.\n\
By default, we use t (\\t) as the delimiter however this may be\n\
specified with the -d option data should be in the following\n\
formats of one entry per line:\n\
printf \"<hostname>\\t<state>\\t<output>\\n\"\n\
printf \"<hostname>\\t<service>\\t<state>\\t<output>\\n\"\n")
        parser.add_argument('-c', '--checktype', action="store",
            dest="checktype", help="1 for passive 0 for active")
        
        options = parser.parse_args()

        if not options.url:
            parser.error('You must specify a url.')
        if not options.token:
            parser.error('You must specify a token.')
        try:
            self.setup(options)
            sys.exit()
        except Exception as e:
            sys.exit(e)
        
    def getText(self, nodelist):
        rc = []
        for node in nodelist:
            if node.nodeType == node.TEXT_NODE:
                rc.append(node.data)
        return ''.join(rc)

    def post_data(self, url, token, xml):
        # Make sure URL ends with a /
        if not url.endswith('/'):
            url += '/'
        params = UrlParse.urlencode({'token': token.strip(), 'cmd': 'submitcheck', 'XMLDATA': xml});
        url_with_data = "{0}?{1}".format(url, params)
        try:
            with UrlRequest.urlopen(url_with_data) as f:
                result = parseString(f.read().decode('utf-8'))
        except Exception as e:
            print("ERROR - Cannot connect to url: {0}".format(str(e)))
            return -3
        if self.getText(result.getElementsByTagName("status")[0].childNodes) == "0":
            return 1
        else:
            print("ERROR - NRDP Returned: " + self.getText(result.getElementsByTagName("message")[0].childNodes))
            return -1

    def setup(self, options):
        if not options.delim:
            options.delim = "\t"
        if not options.checktype:
            options.checktype = "1"
        
        # If only url and token have been provided then it is assumed that data is being piped
        if not options.hostname and not options.state and not options.file:
            xml="<?xml version='1.0'?>\n<checkresults>\n";
            for line in sys.stdin.readlines():
                parts = line.split(options.delim)
                if len(parts) == 4:
                    xml += "<checkresult type='service' checktype='"+options.checktype+"'>"
                    xml += "<hostname>"+html.escape(parts[0],True)+"</hostname>"
                    xml += "<servicename>"+html.escape(parts[1],True)+"</servicename>"
                    xml += "<state>"+parts[2]+"</state>"
                    xml += "<output>"+html.escape(parts[3],True)+"</output>"
                    xml += "</checkresult>"
                if len(parts) == 3:
                    xml += "<checkresult type='host' checktype='"+options.checktype+"'>"
                    xml += "<hostname>"+html.escape(parts[0],True)+"</hostname>"
                    xml += "<state>"+parts[1]+"</state>"
                    xml += "<output>"+html.escape(parts[2],True)+"</output>"
                    xml += "</checkresult>"
                xml += "</checkresults>"
        
        elif options.hostname and options.state:
            xml="<?xml version='1.0'?>\n<checkresults>\n";
            if options.service:
                xml += "<checkresult type='service' checktype='"+options.checktype+"'>"
                xml += "<hostname>"+html.escape(options.hostname,True)+"</hostname>"
                xml += "<servicename>"+html.escape(options.service,True)+"</servicename>"
                xml += "<state>"+options.state+"</state>"
                xml += "<output>"+html.escape(options.output,True)+"</output>"
                xml += "</checkresult>"
            else:
                xml += "<checkresult type='host'  checktype='"+options.checktype+"'>"
                xml += "<hostname>"+html.escape(options.hostname,True)+"</hostname>"
                xml += "<state>"+options.state+"</state>"
                xml += "<output>"+html.escape(options.output,True)+"</output>"
                xml += "</checkresult>"
            xml += "</checkresults>"

        elif options.file:
            try:
                file_handle = open(options.file, 'r')
            except Exception as e:
                print("ERROR - in opening file: " + options.file)
                return -4
            else:
                xml = file_handle.read()
                file_handle.close()

        return self.post_data(options.url, options.token, xml)

if __name__ == "__main__":
    send_nrdp().run()

```

**Comment by xavierba** [comment]
In case this is of any help, here is a commit adding @PHuhn changes from the above comment to current master:
https://github.com/xavierba/nrdp/commit/9a3b430d90e1d5b176cf2542a8317d3beeb5ec5d

I've added a second commit to make the code compatible with both python2 and python3:
https://github.com/xavierba/nrdp/commit/94c66f5e848fa3fdadf2141345ac1977e748cb6c

There is yet another commit in this branch with whitespace fixes:
https://github.com/xavierba/nrdp/commit/17d79ab6dbdbafbae6ba4d29872f12f7a1c48758

I'm not a python developer, so I can't validate this is 100% correct, so not making a PR yet, but it seems to work with both python2 and python3, feel free to review and give feedback, I'll make a proper PR then.

**Comment by xavierba** [comment]
As a side note, I'm maintaining an nrdp rpm package for the Fedora/RHEL ecosystem, and python2 is on the way out, so the python3 compatibility is needed for both upcoming Fedora 31 and the already released RHEL 8, if that gives any incentive to get python3 support merged rather sooner than later ;-)

**Comment by PHuhn** [comment]
Thanks, Xavier.

**Comment by xavierba** [comment]
@PHuhn, short of reaction from Nagios buddies, do you have any comment ?

**Comment by PHuhn** [comment]
I have tried it under python 3 and this is the results:

> pi@raspberrypi:/usr/local/nrdp/clients $ python3 send_nrdp.py
>   File "send_nrdp.py", line 119
>     except Exception, e:
>                     ^
> SyntaxError: invalid syntax
> pi@raspberrypi:/usr/local/nrdp/clients $ 

I can only tell you that Raspian still supports python 2.7 and 3.7.3, but I code in python 3.7 exclusively.

**Comment by xavierba** [comment]
@PHuhn, are you sure you tested the code from my branch ?

**Comment by PHuhn** [comment]
Hey Xavier:
I did not.  I used the following:

    wget -O "nrdp-2.0.2.tar.gz" "https://github.com/NagiosEnterprises/nrdp/archive/2.0.2.tar.gz"

**Comment by PHuhn** [comment]
I just tried the github version and got the follwing:

```
pi@raspberrypi:~/sh/t $ python3 ./send_nrdp.py
Traceback (most recent call last):
  File "./send_nrdp.py", line 50, in <module>
    from future.standard_library import install_aliases
ModuleNotFoundError: No module named 'future'
```

Python version is as follows:

```
pi@raspberrypi:~/sh/t $ python3 --version
Python 3.7.3
```

**Comment by xavierba** [comment]
@jomann09 @hedenface can you please take a look ?
Now that NRDP uses a free software license, I'd really like to move it off from RPM Fusion to Fedora/EPEL, but for that to happen I need python3 support. Hopefully, my work on this support is not too awful and okay to be committed.
Also I hope I've not been wasting my time crafting the patches, but the lack of answer is troubling...

**Comment by xavierba** [comment]
See #45

**Comment by xavierba** [comment]
@PHuhn this is certainly not directed at you, you did the initial porting work, and I'm very grateful to you. What I find troubling is the lack of answer from the nagios buddies. And not only on nrdp front, I've had patches taking dust on the shelf for literally years in nsca, and I'm not the only one. I understand they are likely overworked and I'm definitely not expecting answer within a day or even a week, but months and years passing by are a bit harder to understand.
Sorry for ranting, I guess I just need some rest :-)

**Comment by xavierba** [comment]
@hedenface @jomann09  any chance to look at #45 ?

**Comment by jomann09** [comment]
I haven't merged it in yet because it requires some changes and I haven't had the time right now to make them. The new version of the script should not only work with python 3 it should still work with python 2 since some systems (like CentOS 6/7) still use it and will be using it for a while. That includes changing the top shebang and the imports, etc.

**Comment by xavierba** [comment]
@jomann09 glad to get an answer :-)
The patch makes the script compatible with both python2 and python3. 
Actually, CentOS 6 does have python3 available in the EPEL repository and CentOS 7 does have both python2 and python3 from the base repository since 7.7.
I would be glad to try and fix any comment on the PR. Could you please point out what is lacking, needs changing or whatever over there ?

**Comment by jomann09** [comment]
When the merge was originally posted I tried it locally on a CentOS 6 box and it does not work with both python 2 and 3 without changes. I am aware that you can manually install python 3 but the script should work without that being required. If you take the script that is in the pull request and run it on a clean CentOS 6 system, it should work. Besides the issue with the environment variable at the top, there are imports for modules that don't exist in python 2 and the script will need to deal with that also. We do a similar thing in the check_ncpa.py in the NCPA repo.

**Comment by PHuhn** [comment]
Hi:
I was trying **send_nrdp.py** and it is failing on ***html.escape***.  I am running this on a raspberry pi with the latest buster release.
I tried to install html with the following:
```
pi@raspberrypi:~/sh/t $ sudo python3 -m pip install html
```
It is trying to install the following:
```
https://files.pythonhosted.org/packages/4a/df/0e3d22d50ee43274eb5116f49972a164d853bb3ab305a69a0540b6292252/html-1.16.tar.gz
```
but gets the  following error.
```
    ModuleNotFoundError: No module named 'html.parser'; 'html' is not a package
```
So far I have not found and answer.  So, I am currently stuck.
Phil

**Comment by ericloyd** [comment]
I'm just curious - why not use the bash version of send_nrdp (send_nrdp.sh)?  No Python required and should work on just about any bash-compatible Linux.

**Comment by PHuhn** [comment]
I have the following python3 application:
[py-logger application](https://github.com/PHuhn/py-logger)

**Comment by ericloyd** [comment]
I'm just thinking as a temporary work-around until the python 2/3 is worked out.

import subprocess
subprocess.run(["/usr/local/nrdp/clients/send_nrdp.sh", "arguments])

**Comment by PHuhn** [comment]
Hi Xavier:
I noticed that the version that I created also had html as follows:
```
import argparse, sys, html
```
I had to change the html import to the following:
```
import html
```
and then it worked.
Phil

**Comment by xavierba** [comment]
@PHuhn, thanks for the update.
However, I fell like this won't move until upstream cares, so let's keep waiting on them. But then, python2 is EOL for almost 2 months as I'm writing this.

**Comment by jomann09** [comment]
Merged pull request #45 for python support since Python 2 is EOL. Still should be backwards compatible, but env is set for python3 which I will leave due to the fact that 2 is EOL.
